packages/protobuf/src/create-descriptor-set.ts

```diff
@@ -850,47 +949,77 @@ function isOptionalField(
+  file: DescFile,
   proto: FieldDescriptorProto,
-  syntax: "proto2" | "proto3" | "editions",
-): boolean {
-  assert(proto.type, `invalid FieldDescriptorProto: missing type`);
-  switch (syntax) {
-    case "proto3":
-      switch (proto.type) {
-        case FieldDescriptorProto_Type.DOUBLE:
-        case FieldDescriptorProto_Type.FLOAT:
-        case FieldDescriptorProto_Type.INT64:
-        case FieldDescriptorProto_Type.UINT64:
-        case FieldDescriptorProto_Type.INT32:
-        case FieldDescriptorProto_Type.FIXED64:
-        case FieldDescriptorProto_Type.FIXED32:
-        case FieldDescriptorProto_Type.UINT32:
-        case FieldDescriptorProto_Type.SFIXED32:
-        case FieldDescriptorProto_Type.SFIXED64:
-        case FieldDescriptorProto_Type.SINT32:
-        case FieldDescriptorProto_Type.SINT64:
-        case FieldDescriptorProto_Type.BOOL:
-        case FieldDescriptorProto_Type.ENUM:
-          // From the proto3 language guide:
-          // > In proto3, repeated fields of scalar numeric types are packed by default.
-          // This information is incomplete - according to the conformance tests, BOOL
-          // and ENUM are packed by default as well. This means only STRING and BYTES
-          // are not packed by default, which makes sense because they are length-delimited.
-          return true;
-        default:
-          return false;
-      }
-    case "proto2":
+  resolveFeatures: FeatureResolverFn,
+) {
+  const { repeatedFieldEncoding } = resolveFeatures(file.edition);
+  if (repeatedFieldEncoding != FeatureSet_RepeatedFieldEncoding.PACKED) {
+    return false;
+  }
+  // From the proto3 language guide:
+  // > In proto3, repeated fields of scalar numeric types are packed by default.
+  // This information is incomplete - according to the conformance tests, BOOL
+  // and ENUM are packed by default as well. This means only STRING and BYTES
+  // are not packed by default, which makes sense because they are length-delimited.
+  switch (proto.type) {
+    case FieldDescriptorProto_Type.STRING:
+    case FieldDescriptorProto_Type.BYTES:
+    case FieldDescriptorProto_Type.GROUP:
+    case FieldDescriptorProto_Type.MESSAGE:
       return false;
-    case "editions":
-      // TODO support edition featureset
+    default:
       return true;
   }
 }
 
+/**
+ * Pack this repeated field?
+ *
+ * Respects field type, proto2/proto3 defaults and the `packed` option, or
+ * edition defaults and the edition features.repeated_field_encoding options.
+ */
+function isPackedField(
+  file: DescFile,
+  parent: DescMessage | undefined,
+  proto: FieldDescriptorProto,
+  resolveFeatures: FeatureResolverFn,
+) {
+  switch (proto.type) {
+    case FieldDescriptorProto_Type.STRING:
+    case FieldDescriptorProto_Type.BYTES:
+    case FieldDescriptorProto_Type.GROUP:
+    case FieldDescriptorProto_Type.MESSAGE:
+      // length-delimited types cannot be packed
+      return false;
+    default:
+      switch (file.edition) {
```

Comment:
Agree that it would make sense to check for `LABEL_REPEATED`. Didn't want to change behavior in this PR.

> Also, this default case seems a bit more verbose than I would have expected.

Yes, I wonder how this will play out for other implementations. But the field `packed` of `FieldOptions` is indeed not populated at all with editions.